### DISCUSSION TITLE
How should I draw for pose detection?

### DISCUSSION THREAD

**Discussion by bbhxwl** [ultralytics/ultralytics#19373] [Q&A]
```
import cv2 as cv
from ultralytics import YOLO


model = YOLO("yolo11n-pose.pt")
cap = cv.VideoCapture(0)
while True:
    ret, frame = cap.read()
    if not ret:
        break
    results = model(frame)
    if results:
        for result in results:
            for xy in result.keypoints.xyn:
                # error
                cv.circle(frame, (int(xy[0]), int(xy[1])), 2, (0, 0, 255), -1)

    cv.imshow("frame", frame)
    if cv.waitKey(1) == ord('q'):
        break

# print(y.backward())
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @bbhxwl, thank you for sharing your interest in Ultralytics 🚀! It’s great to see you exploring pose detection with YOLO! 

We recommend checking out the [Docs](https://docs.ultralytics.com/) for helpful resources, particularly the [Tasks](https://docs.ultralytics.com/tasks/) section, which covers pose detection and includes examples that might point you in the right direction.

It looks like you’re encountering an issue in your implementation. If this is a 🐛 bug or an unexpected behavior, please provide more details, such as a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us better understand and debug the issue. Including your dependencies, Python environment, and any error messages would be especially useful.

If you’re exploring ❓ custom usage or have questions about code implementation, feel free to share more context about the task you’re working on. This may include example inputs, outputs, or further clarification.

## Upgrade

Ensure you're using the latest version of the `ultralytics` library and its dependencies. You can update with:

```bash
pip install -U ultralytics
```

Running on an up-to-date platform is often the easiest way to resolve potential issues. YOLO supports running in various verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community Support

Our amazing community is here to help! Join discussions and seek advice in real-time on [Discord](https://discord.com/invite/ultralytics) 🎧. You can also post detailed questions on [Discourse](https://community.ultralytics.com/) or collaborate with others on our [Subreddit](https://reddit.com/r/Ultralytics).

An Ultralytics engineer will assist you soon to help guide you further. Thank you for your patience and for being part of the Ultralytics community! 🚀

**Comment by glenn-jocher** [comment]
@bbhxwl for pose estimation keypoint visualization, you can directly use the `plot()` method from YOLO results. Here's a minimal working example:

```python
from ultralytics import YOLO
import cv2

model = YOLO("yolo11n-pose.pt")
cap = cv2.VideoCapture(0)

while cap.isOpened():
    success, frame = cap.read()
    if not success: break
    
    results = model(frame)
    annotated_frame = results[0].plot()  # Auto-plots keypoints
    
    cv2.imshow("Pose Estimation", annotated_frame)
    if cv2.waitKey(1) == ord('q'): break
```

For custom drawing of keypoints, see the [YOLO11 Pose Estimation documentation](https://docs.ultralytics.com/tasks/pose/) which shows how to access keypoint coordinates in both normalized and pixel formats.

**Reply by bbhxwl** [reply]
> @bbhxwl for pose estimation keypoint visualization, you can directly use the `plot()` method from YOLO results. Here's a minimal working example:
> 
> ```python
> from ultralytics import YOLO
> import cv2
> 
> model = YOLO("yolo11n-pose.pt")
> cap = cv2.VideoCapture(0)
> 
> while cap.isOpened():
>     success, frame = cap.read()
>     if not success: break
>     
>     results = model(frame)
>     annotated_frame = results[0].plot()  # Auto-plots keypoints
>     
>     cv2.imshow("Pose Estimation", annotated_frame)
>     if cv2.waitKey(1) == ord('q'): break
> ```
> 
> For custom drawing of keypoints, see the [YOLO11 Pose Estimation documentation](https://docs.ultralytics.com/tasks/pose/) which shows how to access keypoint coordinates in both normalized and pixel formats.

Thank you, I have learned YOLO11's object detection, it is really powerful. I know he can use labelimg annotation, but he doesn't know how to self train YOLO11's pose yet. Which document should I refer to for learning? Mainly, I don't know what tool to use for annotation, and what should be written in datayaml format

**Reply by glenn-jocher** [reply]
@bbhxwl for training YOLO11-pose models, refer to the Pose Estimation documentation which details dataset preparation using tools like Ultralytics HUB for annotation. The Dataset Format section explains how to structure your data.yaml and convert existing annotations with the JSON2YOLO converter. A minimal training example would be:

```python
from ultralytics import YOLO
model = YOLO("yolo11n-pose.pt")
model.train(data="your-data.yaml", epochs=100)
```

**Reply by bbhxwl** [reply]
> @bbhxwl for training YOLO11-pose models, refer to the Pose Estimation documentation which details dataset preparation using tools like Ultralytics HUB for annotation. The Dataset Format section explains how to structure your data.yaml and convert existing annotations with the JSON2YOLO converter. A minimal training example would be:
> 
> ```python
> from ultralytics import YOLO
> model = YOLO("yolo11n-pose.pt")
> model.train(data="your-data.yaml", epochs=100)
> ```
> 
> Commercial use requires an Ultralytics Enterprise License.

Are you saying that Ultralytics HUB business requires a fee, or does YOLO11 require a fee for any business?

**Reply by bbhxwl** [reply]
> @bbhxwl for commercial use of any Ultralytics software (including YOLO11 models and Ultralytics HUB), an Enterprise License is required unless your entire project is open-sourced under AGPL-3.0. The Ultralytics Licensing page provides full details on usage scenarios and requirements. Ultralytics HUB offers free tiers for AGPL-compliant projects and Enterprise tiers for commercial use.

What is the general cost of authorization? Do you know?

**Reply by glenn-jocher** [reply]
For Enterprise License pricing tailored to your organization's size and use case, please visit the Ultralytics Licensing page to request a custom quote. Costs vary depending on deployment scope and commercial requirements.
